bpfman/src/multiprog/tc.rs

```diff
@@ -428,26 +434,50 @@ impl TcDispatcher {
                     Direction::Ingress => TcAttachType::Ingress,
                     Direction::Egress => TcAttachType::Egress,
                 };
-                if let Ok(old_link) =
-                    SchedClassifierLink::attached(&if_name, attach_type, priority, old_handle)
-                {
-                    let detach_result = old_link.detach();
-                    match detach_result {
-                        Ok(_) => debug!(
-                            "TC dispatcher {}, {}, {}, {} successfully detached",
-                            if_name, direction, priority, old_handle
-                        ),
-                        Err(_) => debug!(
-                            "TC dispatcher {}, {}, {}, {} not attached when detach attempted",
-                            if_name, direction, priority, old_handle
-                        ),
-                    }
-                }
+                if let Some(netns) = netns {
+                    if let Ok(_netns_guard) = enter_netns(netns) {
+                        self.detach_dispatcher(
+                            &if_name,
+                            attach_type,
+                            priority,
+                            old_handle,
+                            direction,
+                        );
+                    } else {
+                        debug!("enter_netns failed.  The netns may have been deleted.");
+                    };
+                } else {
+                    self.detach_dispatcher(&if_name, attach_type, priority, old_handle, direction);
+                };
             };
         }
         Ok(())
     }
 
+    fn detach_dispatcher(
+        &mut self,
+        if_name: &str,
+        attach_type: TcAttachType,
+        priority: u16,
+        handle: u32,
+        direction: Direction,
+    ) {
+        if let Ok(old_link) = SchedClassifierLink::attached(if_name, attach_type, priority, handle)
+        {
+            let detach_result = old_link.detach();
+            match detach_result {
+                Ok(_) => debug!(
+                    "TC dispatcher {}, {}, {}, {} successfully detached",
+                    if_name, direction, priority, handle
+                ),
+                Err(_) => debug!(
+                    "TC dispatcher {}, {}, {}, {} not attached when detach attempted",
+                    if_name, direction, priority, handle
+                ),
+            }
+        }
+    }
```

Comment:
what about doing the same for TCX and XDP programs ?

Reply:
We don't have this particular problem with TCX, XDP and Uprobe programs becasue simply deleting the `pin` files causes the programs to be deleted.  TC is different because we have to manage the detaching ourselves when the dispatcher is no longer needed.

Reply:
And, you can delete the pin files without entering the network namespace.